pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV2.java

```diff
@@ -200,39 +138,40 @@ public DataTableImplV2(ByteBuffer byteBuffer)
       Map<String, String> metadata = new HashMap<>(numEntries);
 
       for (int i = 0; i < numEntries; i++) {
-        String key = decodeString(dataInputStream);
-        String value = decodeString(dataInputStream);
+        String key = DataTableUtils.decodeString(dataInputStream);
+        String value = DataTableUtils.decodeString(dataInputStream);
         metadata.put(key, value);
       }
 
       return metadata;
     }
   }
 
-  private static String decodeString(DataInputStream dataInputStream)
-      throws IOException {
-    int length = dataInputStream.readInt();
-    if (length == 0) {
-      return StringUtils.EMPTY;
-    } else {
-      byte[] buffer = new byte[length];
-      int numBytesRead = dataInputStream.read(buffer);
-      assert numBytesRead == length;
-      return StringUtil.decodeUtf8(buffer);
-    }
-  }
-
   @Override
   public void addException(ProcessingException processingException) {
     _metadata.put(EXCEPTION_METADATA_KEY + processingException.getErrorCode(), processingException.getMessage());
   }
 
+  // getExceptions return a map of errorCode->errMessage of the datatable.
+  @Override
+  public Map<Integer, String> getExceptions() {
+    Map<Integer, String> exceptions = new HashMap<>();
+    for (String key : _metadata.keySet()) {
+      if (key.startsWith(EXCEPTION_METADATA_KEY)) {
+        // In V2, all exceptions are added into metadata, using "Exception"+errCode as key,
+        // Integer.parseInt(key.substring(9)) can extract the error code from the key.
+        exceptions.put(Integer.parseInt(key.substring(9)), _metadata.get(key));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.